Er 14 as implement api equipment removal

amansiid opened the pull request:
/**
 *  DELETE/EQUIPMENT REMOVE SINGLE/MULTIPLE
 *  URL => /api/inventory/remove
 * 
 *      "schoolId": string,
 *      "serialId": ["serialNum", "serialNum", ...]
 * 
 *  @return 400 (If failed validation)
 *          503 (If server error) 
 *          200 (Delete Successful)
 */

This API allows the removal of items from the inventory, single items or multiple. 
The API should return a response message with status 200, indicates the action of removal was successful or return an error message with status 503.

amansiid committed:
setting up

amansiid committed:
updates

amansiid committed:
DeleteItem for multiple and single

API functioning but not displaying message 200

amansiid committed:
display is 200 fixed

amansiid committed:
Merge branch 'main' into ER-14-AS-Implement-API-EquipmentRemoval

tranleduyan commented on src_code/src/routes/InventoryRoutes.js:
```diff
@@ -1,485 +1,927 @@
+})
+
+/** Exports the router */
+module.exports = router;
```
There is a duplicate.

YiJiePV replied:
Yep, what Le said. When resolving merge conflicts, don't forget to double check for duplicate code and bad syntax (like the `===` divider) before commiting and pushing the changes 😉 And while at it, don't forget to leave a newline at the end of the files

tranleduyan reviewed (changes requested):
After all the changes, you can re test it again.

YiJiePV commented on src_code/src/controllers/services/Inventory/EquipmentRemoval.js:
```diff
@@ -4,25 +4,168 @@ const db = require("../../../configurations/database/DatabaseConfigurations");
+ * Validates the request parameters for single equipment removal.
+ *
+ * @param {Object} res - Express response object.
+ * @param {Object} req - Express request object containing item ID
```
nit: and school ID??

YiJiePV commented on src_code/src/controllers/services/Inventory/EquipmentRemoval.js:
```diff
@@ -4,25 +4,168 @@ const db = require("../../../configurations/database/DatabaseConfigurations");
+}
+
+/**
+ * Validates the request parameters for single equipment removal.
```
nit: this also has the capability for mass delete, right? Update?

YiJiePV commented on src_code/src/controllers/services/Inventory/EquipmentRemoval.js:
```diff
@@ -4,25 +4,168 @@ const db = require("../../../configurations/database/DatabaseConfigurations");
 module.exports = {
-    EquipmentRemoval
+    EquipmentRemoval,
+    EquipmentRemovalValidation
```
Do we really need to export the validation here?

YiJiePV commented on src_code/src/routes/InventoryRoutes.js:
```diff
@@ -1,485 +1,927 @@
+ *  URL => /api/inventory/remove
+ * 
+ *      "schoolId": string,
+ *      "itemId": serial
```
This is the request body, right? And any info about the response?

YiJiePV commented on src_code/src/utils/services/helper_functions/db_helper_functions/item/ItemDBHelperFunction.js:
Does this helper serve a purpose? If not, I think it would be wise to delete it.

YiJiePV reviewed (changes requested):
Looking good! Just a couple house-keeping things from my end, and Le's feedback as well. Unfortunately, I cannot test your code as of now, since the error pointed out by Le and myself is causing the error:
![syntax-error](https://github.com/tranleduyan/ETC-Backend/assets/97055517/0168a229-7ef3-4e74-9d51-8eb73a56f2f2)
Once that error's fixed, I'll be able to test your code and make sure that everything works smoothly 👍

amansiid committed:
The issues are resolved

amansiid commented:
I've addressed the requested updates.

leduyantran commented on src_code/src/routes/InventoryRoutes.js:
```diff
@@ -501,13 +62,17 @@ router.post('/equipment', async(request, response) => {
+ *          200 (Delete Successful)
  */
-router.delete('/equipment', async(request, response) => {
+router.delete('/remove', async(request, response) => {
```
Why is this change to remove? If you look at the code base logic, all post, delete are in one url. There is no sub endpoint or necessary to create another url. Also, if there is the inventory/remove endpoint does not meaningful to say it is for deleting equipment. The url should be /equipment.  Please update the comment as well.

leduyantran committed:
Update route

leduyantran committed:
Clean up and handle rollback. Optimize code.

If no serialId, then it is mean no item deleted, then just return 200 instead (no errors) but if missing schoolid is a big problem (see line 70) this save more time and you dont have to handle line 37 to line 40 with await.

line 117 use items?.length because it is possible that non of the serial id is an exist item therefore it will return null then null cannot be access with length

line 51, we need to roll back the transaction also to maintain the database stay the same if delete process has error

move const trx to outside of try catch block so that it can be use globally in the function equipment removal (line 31)

leduyantran reviewed (approved):
The code is looking good to me. I tested and it works well.

leduyantran committed:
Resolving conflicts

leduyantran committed:
Merge branch 'main' into ER-14-AS-Implement-API-EquipmentRemoval

leduyantran committed:
Update comments

mohammedking11-spu commented on src_code/src/controllers/services/Inventory/EquipmentRemoval.js:
Looks easy to follow. I appreciate that you included a comment that shows what the expected request body would look like. It was well-structured, and enough comments to make everything clear and concise. Great job at sticking with the code conventions.

YiJiePV committed:
fixed BadRequest method call

mohammedking11-spu reviewed (approved):
Overall, it gets my approval. It was easy to follow and well-structured, and the good thing is that it also has enough comments. Looks optimized and works as requested! Great job Aman!

YiJiePV committed:
Added to comment

YiJiePV reviewed (approved):
Much better now! Everything's working like they should, I just added a couple minor fixes myself. I'm a little concerned about the fact that there isn't a check for each individual serial ID passed in the array, but since a "item not found" 400 Bad Request is thrown every time I tried that test case, I think it would probably do for now. But besides that, no other issues from my end 👍
